Update README (#1253)
memorysafety/rav1d: README.md

# rav1d

**rav1d** is an AV1 cross-platform decoder, open-source, and focused on speed
and correctness. It is a Rust port of
[dav1d](https://code.videolan.org/videolan/dav1d).

# Building

rav1d is written in Rust and uses the standard Rust toolchain to build. The Rust
toolchain can be installed by going to https://rustup.rs. The rav1d library
builds on stable Rust for x86, x86_64, and aarch64, but currently requires a

> nightly compiler for armv7 and riscv. The project is configured to use a nightly

Reviewer: Should we mention why?  Since it's a minor issue.

Also, I'm not sure if it's important to be pedantic here, but if we're using `target_arch` names here, then it's just `arm`, not `armv7`.  If we're using `--target` names, then it is `armv7`, but it's `i686` instead of `x86`.  For whichever we use, I think it's also useful to say if we mean `#[cfg(target_arch = "")]`s or `--target`s.

Also, I think we should put these in `` `backticks` ``, in order to contrast with non-specific usages like x86 below, where we really mean `x86` or `x86_64`.

Reviewer: Actually I think the `target_arch` names are better, since they're the ones we're actually detecting.  For example, I think `riscv32*` targets should also work since we support `any(target_arch = "riscv32", target_arch = "riscv64")`.

Author: Changed to use the `target_arch` names, but I'm leaving riscv32 out because we don't yet check it in CI, so it's not yet a fully supported platform.

compiler by default via `rust-toolchain.toml`, but a stable build can be made

> with the `+stable` cargo flag.

Reviewer: You also need to only build the library with `--lib`.  By default `cargo build` will try to build the other binaries, too, which still use a bunch of nightly features.

Reviewer: Oh I see you put the `--lib` below.  Maybe just add here "a stable `--lib` build" or "a stable build of the library".

For x86 targets, you'll also need to install [`nasm`](https://nasm.us/) in order
to build with assembly support.

A release build can then be made using cargo:

```sh
cargo build --release
```

For development purposes you may also want to use the `opt-dev` profile, which
runs faster than a regular debug build but has all debug checks still enabled:

```sh
cargo build --profile opt-dev
```

To build just librav1d using a stable compiler:

```sh
cargo +stable build --lib --release
```

## Feature Flags

The following feature flags are supported:

* `asm` - Enables optimized assembly routines, if available for the target
  platform.
* `bitdepth_8` - Enables support for 8 bitdepth decoding.
* `bitdepth_16` - Enables support for 10 and 12 bitdepth decoding.

All of these features are enabled by default. In order to build a version of

> librav1d that disables one or more of these features use the

Reviewer: ```suggestion
`librav1d` that disables one or more of these features use the
```

`--no-default-features` flag in combination with the `--features` flag to enable
any desired features. For example, to build without assembly routines, which is
useful when testing the Rust fallback functions, do the following:

```sh
cargo build --no-default-features --features="bitdepth_8,bitdepth_16"
```

> ## Cross-Compiling

Reviewer 2: We should list the supported targets and make a note that targets other than host need to be installed, e.g. with rustup.

Author: Added a list of supported targets and updated the wording to note that the target will have to be installed via rustup.

rav1d can be cross-compiled for a target other than the host platform using the

> Cargo `--target` flag. This will require passing additional arguments to the
> Rust compiler to tell it what linker to use. This can be done by setting the

Reviewer: ```suggestion
`cargo` `--target` flag. This will require passing additional arguments to
`rustc` to tell it what linker to use. This can be done by setting the
```

`RUSTFLAGS` enviroment variable and specifying the `linker` compiler flag. For

> example, compiling for `aarch64-unknown-linux-gnu` from a Ubuntu Linux machine

Reviewer: ```suggestion
example, compiling for `aarch64-unknown-linux-gnu` from an Ubuntu Linux machine
```

would be done as follows:

```sh
RUSTFLAGS="-C linker=aarch64-linux-gnu-gcc" cargo build --target aarch64-unknown-linux-gnu
```

> If you're cross-compiling in order to run tests under QEMU you'll also need to
> specify the `+crt-static` target feature.

Reviewer: ```suggestion
If you're cross-compiling in order to run tests under QEMU (`qemu-*-static`),
you'll also need to specify the `+crt-static` target feature.
```

I think we actually only need it for static `qemu` builds.

```sh
RUSTFLAGS="-C target-feature=+crt-static -C linker=aarch64-linux-gnu-gcc" cargo build --target aarch64-unknown-linux-gnu
```

This will require installing the appropriate cross-platform compiler and linker
toolchain for your target platform. Examples of how we cross-compile rav1d in CI

> can be found in `.github/workflows/build-and-test-qemu.yml`.

Reviewer: ```suggestion
can be found in [`.github/workflows/build-and-test-qemu.yml`](.github/workflows/build-and-test-qemu.yml).
```

## Running Tests

Currently we use the original [Meson](https://mesonbuild.com/) test suite for
testing the Rust port. This means you'll need to [have Meson
installed](https://mesonbuild.com/Getting-meson.html) to run tests.

To setup and run the tests, do the following:

> First, build the Rust project using Cargo. You'll need to do this step manually
> before running any tests because it is not built automatically when tests are
> run. It's recommended to run tests with either the release or opt-dev profile as
> the debug build runs slowly and often causes tests to timeout. The opt-dev

Reviewer: ```suggestion
First, build `rav1d` using `cargo`. You'll need to do this step manually
before running any tests because it is not built automatically when tests are
run. It's recommended to run tests with either the `release` or `opt-dev` profile as
the debug build runs slowly and often causes tests to timeout. The `opt-dev`
```

profile is generally ideal for development purposes as it enables some
optimizations while leaving debug checks enabled.

```sh
cargo build --release
```

Or:

```sh
cargo build --profile opt-dev
```

> Then you can run the tests with the `tests.sh` helper script:

Reviewer: ```suggestion
Then you can run the tests with the [`test.sh`](.github/workflows/test.sh) helper script:
```

```sh
.github/workflows/test.sh -r target/release/dav1d
```

Or:

```sh
.github/workflows/test.sh -r target/opt-dev/dav1d
```

The test script accepts additional arguments to configure how tests are run:

> * `-s PATH` - Specify a path to the seek-stress binary in order to run the
>   seek-stress tests. This is generally in the same output directory as the main

Reviewer: ```suggestion
* `-s PATH` - Specify a path to the `seek-stress` binary in order to run the
  `seek-stress` tests. This is generally in the same output directory as the main
```

Reviewer: Should this be included in the default test command?  We pretty much always run it with `seek_stress`.

Author: I usually run without the seek stress tests. I don't think we need to add it here.

Reviewer: Oh, okay then. I always do.

Reviewer: Can you add the backticks, though?

  `dav1d` binary, e.g. `target/release/seek_stress`.
* `-t MULTIPLIER` - Specify a multiplier for the test timeout. Allows for tests
  to take longer to run, e.g. if running tests with a debug build.
* `-f DELAY` - Specify a frame delay for the tests. If specified the tests will
  also be run with multiple threads.
* `-n` - Test with negative strides.
* `-w WRAPPER` - Specify a wrapper binary to use to run the tests. This is
  necessary for testing under QEMU for platforms other than the host platform.

You can learn more about how to build and test by referencing the CI scripts in

> the `.github/workflows` folder.

Reviewer: ```suggestion
the [`.github/workflows`](.github/workflows) folder.
```

# Using rav1d

> rav1d is designed to be a drop-in replacement for dav1d, so it primarily exposes
> a C API with the same usage as dav1d's. This is found in the `librav1d.a`
> library generated by `cargo build`. dav1d's primary API documentation [can be

Reviewer: ```suggestion
`librav1d` is designed to be a drop-in replacement for `libdav1d`, so it primarily exposes
a C API with the same usage as `libdav1d`'s. This is found in the `librav1d.a`
library generated by `cargo build`. `libdav1d`'s primary API documentation [can be
```

I think it's important to be clear about the lib vs. binary difference.  Our code, `librav1d`, is a drop-in replacement for `libdav1d`, and we did not port the `dav1d` (or `seek_stress`) binary to Rust beyond the `c2rust` initial transpilation.  I think it's also important to clarify that, as people will wonder what that code is and why it's not cleaned up or safe at all.

> library generated by `cargo build`. dav1d's primary API documentation [can be
> found here](https://videolan.videolan.me/dav1d/dav1d_8h.html) for reference, and

Reviewer: ```suggestion
library generated by `cargo build`. dav1d's [primary API documentation can be
found here](https://videolan.videolan.me/dav1d/dav1d_8h.html) for reference, and
```

It's important to include what the link is about in the link text for screen readers.  For example, "can be found here" isn't a very useful description.

> the equivalent Rust functions can be found in `src/lib.rs`. You can also
> reference the dav1d binary's code to see how it uses the API, which can be found
> at `tools/dav1d.rs`.

Reviewer: ```suggestion
the equivalent Rust functions can be found in [`src/lib.rs`](src/lib.rs). You can also
reference the `dav1d` binary's code to see how it uses the API, which can be found
at [`tools/dav1d.rs`](tools/dav1d.rs).
```

platform. Examples of how we cross-compile rav1d in CI can be found in
[`.github/workflows/build-and-test-qemu.yml`](.github/workflows/build-and-test-qemu.yml).

The following targets are currently supported:

* `x86_64-unknown-linux-gnu`
* `i686-unknown-linux-gnu`
* `armv7-unknown-linux-gnueabihf`
* `aarch64-unknown-linux-gnu`
* `riscv64gc-unknown-linux-gnu`

## Running Tests

Currently we use the original [Meson](https://mesonbuild.com/) test suite for
testing the Rust port. This means you'll need to [have Meson
installed](https://mesonbuild.com/Getting-meson.html) to run tests.

To setup and run the tests, do the following:

First, build `rav1d` using `cargo`. You'll need to do this step manually before
running any tests because it is not built automatically when tests are run. It's
recommended to run tests with either the `release` or `opt-dev` profile as the
debug build runs slowly and often causes tests to timeout. The `opt-dev` profile
is generally ideal for development purposes as it enables some optimizations
while leaving debug checks enabled.

```sh
cargo build --release
```

Or:

```sh
cargo build --profile opt-dev
```

> Then you can run the tests with the [`tests.sh`](.github/workflows/test.sh)

Reviewer: ```suggestion
Then you can run the tests with the [`test.sh`](.github/workflows/test.sh)
```

# rav1d

**rav1d** is an AV1 cross-platform decoder, open-source, and focused on speed and correctness. It is a Rust port of [dav1d](https://code.videolan.org/videolan/dav1d).

# Building

rav1d is written in Rust and uses the standard Rust toolchain to build. The Rust toolchain can be installed by going to https://rustup.rs. rav1d currently builds with a nightly compiler, which is specified in `rust-toolchain.toml`. The correct nightly build will automatically be installed for you when you run `cargo build`.

> For x86 targets you'll also need to install [nasm](https://nasm.us/) in order to build with assembly support.

Reviewer: ```suggestion
For x86 targets, you'll also need to install [`nasm`](https://nasm.us/) in order to build with assembly support.
```

A release build can then be made using cargo:

```txt
cargo build --release
```

> For development purposes you may also want to use the opt-dev profile, which runs faster than a regular debug build but has all debug checks still enabled:

Reviewer: ```suggestion
For development purposes you may also want to use the `opt-dev` profile,
which runs faster than a regular debug build (which can be quite slow) but has all debug checks still enabled:
```

> ```txt

Reviewer: ```suggestion
```sh
```

cargo build --profile opt-dev
```

## Cross-Compiling

> rav1d can be cross-compiled for a target other than the host platform using the Cargo `--target` flag. This may require passing additional arguments to the Rust compiler to tell it what linker to use. This can be done by setting the `RUSTFLAGS` enviroment variable and specifying the `linker` compiler flag. You'll also need to specify the `+crt-static` target feature. For example, compiling for `aarch64-unknown-linux-gnu` from a Ubuntu Linux machine would be done as follows:

Reviewer: I thought `+crt-static` is only needed when you want to run the cross-compiled binary with `qemu-user-static`.

Author: Ah that's true, I've made a note about that.

> ```txt

Reviewer: ```suggestion
```sh
```

Reviewer: Same applies to all of the other `txt`s.  They can be `sh` if it's directly a shell command.  If it's shell input and output, with a `> ` or `$ ` and then the output, you can use `shell` as the language.

* `-t MULTIPLIER` - Specify a multiplier for the test timeout. Allows for tests to take longer to run, e.g. if running tests with a debug build.
* `-f DELAY` - Specify a frame delay for the tests. If specified the tests will also be run with multiple threads.
* `-n` - Test with negative strides.
* `-w WRAPPER` - Specify a wrapper binary to use to run the tests. This is necessary for testing under QEMU for platforms other than the host platform.

You can learn more about how to build and test by referencing the CI scripts in the `.github/workflows` folder.

# Using rav1d

rav1d is designed to be a drop-in replacement for dav1d, so it primarily exposes a C API with the same usage as dav1d's. This is found in the `librav1d.a` library generated by `cargo build`. dav1d's primary API documentation [can be found here](https://videolan.videolan.me/dav1d/dav1d_8h.html) for reference, and the equivalent Rust functions can be found in `src/lib.rs`. You can also reference the dav1d binary's code to see how it uses the API, which can be found at `tools/dav1d.rs`.

> A Rust API [is planned](https://github.com/memorysafety/rav1d/issues/1252) for addition in the future.

Reviewer: ```suggestion
A [Rust API is planned](https://github.com/memorysafety/rav1d/issues/1252) for addition in the future.
```

RUSTFLAGS="-C target-feature=+crt-static -C linker=aarch64-linux-gnu-gcc" cargo build --target aarch64-unknown-linux-gnu
```

This will require installing the `rustup` component for the target platform and
the appropriate cross-platform compiler/linker toolchain for your target
platform. Examples of how we cross-compile rav1d in CI can be found in
[`.github/workflows/build-and-test-qemu.yml`](.github/workflows/build-and-test-qemu.yml).

The following targets are currently supported:

* `x86_64-unknown-linux-gnu`
* `i686-unknown-linux-gnu`
* `armv7-unknown-linux-gnueabihf`
* `aarch64-unknown-linux-gnu`

> * `riscv64gc-unknown-linux-gnu`

Reviewer: Do we support any `target_arch = "riscv32"` targets?  Because we have `#[cfg]`s for that `target_arch`, but I'm not sure which targets it's meant to correspond to.

Author: We don't test it in CI so I don't think we want to list it here for now.

Reviewer: Yeah, I saw https://github.com/memorysafety/rav1d/pull/1182#discussion_r1663741892 about it, too; makes sense.
